Assault slash changes (for review and eventual pull)

LordXenon opened the pull request:
Throwing a pull request up for initial review by exec. Assault Slash is currently missing training, so that's next, but yeah.

exectails commented on src/ChannelServer/Network/Sending/Send.Skills.cs:
```diff
@@ -665,5 +695,23 @@ public static void CollectAnimationCancel(Creature creature)
+		/// <param name="creature"></param>
+		/// <param name="fPos"></param>
+		/// <param name="tPos"></param>
+		public static void AssaultSlashJump(Creature creature, Position fPos, Position tPos)
```
Please use descriptive variable names or document them in the param comment.

exectails commented on src/ChannelServer/Skills/Combat/AssaultSlash.cs:
```diff
@@ -0,0 +1,141 @@
+			var unk1 = packet.GetString();
+			var targetEntityId = packet.GetLong();
+ 
+			if (attacker.Target != null)
```
If you do the following you have one less indention. Because of the return you don't need the else anymore, which makes the code a little easier to read.

```
if (attacker.Target == null)
{
	Send.Notice(attacker, Localization.Get("Select a target first."));
	Send.SkillPrepareSilentCancel(attacker, skill.Info.Id);
	return false;
}
```

exectails commented on src/ChannelServer/Skills/Combat/AssaultSlash.cs:
```diff
@@ -0,0 +1,141 @@
+				var target = attacker.Region.GetCreature(targetEntityId);
+				var tPos = attacker.GetPosition().GetRelative(target.GetPosition(), -10);
+
+				if (target.IsUnstable && target.RaceData.Is(RaceStands.KnockDownable))
```
`target.IsKnockedDown` would probably be better. IsUnstable checks for Stability <= 0, but that's over quickly and has nothing to do with being knocked down.

exectails commented on src/ChannelServer/Skills/Combat/AssaultSlash.cs:
```diff
@@ -0,0 +1,141 @@
+					Send.SkillUse(attacker, skill.Info.Id, target.EntityId);
+				}
+				
+				else if (!attacker.GetPosition().InRange(target.GetPosition(), attacker.AttackRangeFor(target) + (int)skill.RankData.Var3))
```
Same as above, you can do these two checks in ifs before you come to the success. This is not a must by any means, just a suggestion. Personally I find it much easier to read and follow.

```
// Check this
if(x)
{
	// fail
	return;
}

// Check that
if(y)
{
	// fail
	return;
}

// Passed all checks? Yay!
// success
```

exectails commented on src/ChannelServer/Skills/Combat/AssaultSlash.cs:
```diff
@@ -0,0 +1,141 @@
+			var range = skill.RankData.Var3;
+			var target = creature.Region.GetCreature(targetEntityId);
+
+			Send.Effect(creature, Effect.AssasultSlash, (byte)4, (long)targetEntityId);
```
You only need a cast (`(long)`) if the type isn't clear. `4` is an int by default, but it has to be a byte, so you cast it. But a variable that is a long is a long, and doesn't need casting. More oversight than problem though.

exectails commented on src/ChannelServer/Skills/Combat/AssaultSlash.cs:
```diff
@@ -0,0 +1,141 @@
+			tAction.Stun = StunTime;
+			target.Stability = Creature.MinStability;
+
+			if (!tAction.Has(TargetOptions.Critical))
```
I have a log here with a knock down without critical, are there other factors?

exectails replied:
Also, no shove? do they bounce when you knock them down again?

LordXenon replied:
Yes, by default AS knocks them down again. And under other circumstances it knocks them back. I figured perhaps a crit. But I set it to knockdown when it doesn't crit, and to knock them back if it does crit o.o

DeleteIT replied:
Semi related, can we refer to the two events differently?

Like, knockdown & blowback, or something? I can't be the only person misreading these on a semi-regular basis.

DeleteIT commented on src/ChannelServer/Network/Sending/Send.Skills.cs:
```diff
@@ -665,5 +695,23 @@ public static void CollectAnimationCancel(Creature creature)
+		/// <param name="fPos"></param>
+		/// <param name="tPos"></param>
+		public static void AssaultSlashJump(Creature creature, Position fPos, Position tPos)
+		{
```
Shouldn't this be in the skill itself, since you don't actually do this anywhere else?
At least in Send.Misc?

exectails replied:
No, every single send method goes into Send. It's just a convention to have everything in one place. Also, we might replace Send at some point, with a slightly different method, which would become complicated if there are random packets being built everywhere.

DeleteIT replied:
So, should I have put the taming related things in the Send.Skill segment of the class?

I chose Send.Misc because they didn't fit with the overall "theme" of the class, which was directly related to all skills being activated/processed.

LordXenon commented:
Fixed some things listed; added a new method regarding stability to determine whether to knock down the enemy (i.e. bounce them) or blow them back. Either way, we reset them to the min stability afterwards because AS is a dick like that. But not before, because that would.. break the check!

DeleteIT commented:
>Assault Slash is currently missing training, so that's next

When is that coming, by the way?

exectails commented:
What's the status on this?
